packages/internal/src/utils.js

```diff
@@ -411,37 +411,50 @@ export const assertAllDefined = obj => {
   }
 };
 
-const neverDone = harden({ done: false, value: null });
+/** @type {IteratorResult<undefined, never>} */
+const neverDone = harden({ done: false, value: undefined });
+
+/** @type {IteratorResult<never, void>} */
+const alwaysDone = harden({ done: true, value: undefined });
```

Comment:
I love seeing the two placements of `never` in these two type declarations. Helps me understand!

I am happy that you got rid of the `null`s and replaced them with `undefined`s. I was puzzled earlier when I saw `null` and should have asked you about it then.

What's the difference between the types `undefined` vs `void`, and why is each the right choice for its use above?

Reply:
`undefined` means "you can depend on this value to be `undefined`", but `void` means "you should not depend on the type of this value".  In the case of `IteratorResult<T, TReturn = any> = IteratorYieldResult<T> | IteratorReturnResult<TReturn>`, it's declaring that the return result's value should be ignored by the caller.